:wrench: chore: Implementação de Validação

eduardortm opened the pull request:
Validação de usuario logado

Validação se houve alteração do usuario

Validação do email ou cpf já cadastrado

implementação da rota de update de usuário

eduardortm committed:
:wrench: chore: Implementação de Validação

Validação de usuario logado

Validação se houve alteração do usuario

Validação do email ou cpf já cadastrado

implementação da rota de update de usuário

eduardortm committed:
:wrench: chore; Criação do cadastro de cliente

Criação do middleware

Criação do schemaCustomer

Alteração da validação de existência

gabrielarrupes commented:
Oiii, ao analisar o PR observei algumas coisas que seria interessante alterarmos para seguir padrões e boas práticas.

Segue:

**- INDEX.JS**
`require("dotenv").config();`  - seria interessante colocar no início da página, pois ele não é declarado com const, para padronizar.

`app.listen(process.env.PORT)` optar por `app.listen(process.env.PORT || 3000);`
para seguir o padrão e evitar erros caso alguma eventualidade com process.env.PORT


**EXISTSINDATABASE.js** 

- função deve ser declarada com const;
- Acredito que seria interessante usar um return nos ifs, para não pular para o próximo e mostrar a mensagem incorreta.
 Não sei se no caso abaixo o throw new error retornaria os dois erros juntos, mas acredito que nao.
Ex: se o email já existe e o CPF também, só retornará "CPF já cadastrado" e não sobre o email. E na ordem do front-end o email.

E como o emailExistsInDatabase é usado para cadastrar o usuário (onde não é pedido CPF) colocar um return no if(emailExistsInDatabase) já retornaria o erro direto, sem gastar processamento validando cpf.


- `if (emailExistsInDatabase) {
      throw new Error("Email já cadastrado");
    }` 
- `if (cpfExistsInDatabase) {
      throw new Error("CPF já cadastrado");
    }` 


    
**ROUTER – router.js**

`router = express();` utilizar "const" para declarar


`router.post(“/Login”`) – prefira `router.post(“/login”)` para seguir o padrão dos outros endpoints;


**SCHEMAS**

** Message: **
deve informar quais os erros que podem dar e apartir daí enviar mensagens de erro. Exemplo, se retorna um erro proveniente de required( **mesmo que a validação exista no front**) e esse erro chegue no backend ele retornará uma mensagem em inglês padrão da api do joi.
Por isso devemos validar todos os possíveis erros do backend


**CostumerSchema.js**

**Name: joi**

* alidar mensagens de erro provenientes da string(), pattern() e required() em message:
 
> Alguns possiveis erros: string.base, string.empty,  string.pattern.base, any.required
 
    
**Email:joi**
*	Validar mensagens de erro provenientes da string(), pattern(), email() e required() em message:
	
>       "any.required": 
>       "string.email": 
>       "string.base":
>       
**
telephone.joi:**
*    Validar mensagens de erro provenientes da strings(), pattern() em message()

**cep: joi**

 - Seria interessante usar a regex de apenas números e definir um tamanho máximo de 8 caracteres.
 
**Estado:** Seria interessante definir max dois carateres, e pattern de somente letras.

**cep, logradouro, complemento, bairro, cidade, estado**: validar mensagens de erro;

- Na doc consta todos os possiveis erros que os metodos do joi retornam


**UserSchema**

Em nome, email e telefone proceder validar as mensagens de erro que nem no CostumerSchema;


* Password 
   *  Acredito que seria interessante usar um único pattern() usando uma regex que em uma senha faz todas as validações
        Ex: regex(
          /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/
        )
>     "O campo senha deve conter ao menos 8 caracteres, contendo pelo menos um caractere maiúsculo, minúsculo, numeral e também conter um caractere especial",

   (Pode reformular essa frase se quiser kkk, mas seguindo o padrão da Regex);
   

*  Validar os erros proveniente de string, required, min e pattern;

**UserUpdateSchema.js**

* Tratar os erros também, neste caso só ignorar o que falei sobre required()


**MIDDLEWARES**

**CustomerRegistrationValidatio.js**

Em (database – “costumer”) acho que seria interessante usar `const database = “costumer”)` e colocar essa variável como parâmetro ao invés de declará-la ali.

**UserLoggerValidation**


`const id = jwt.decode(token, process.env.JWT_PASS);`


 Acredito que seria interessante utilizar o método verify() ao invés de decode(), pois é mais seguro para verificar a autenticidade do token. 
Encontrei essa postagem que explica um pouco sobre:  https://community.auth0.com/t/does-jwt-decode-functionality-verify-the-token-or-only-decode-the-token-if-yes-what-all-does-the-decode-function-do/9509


**UserUpdateValidation**

Função **areObjectsEqual**

* seria interessante seguir o padrão de arrow function como está nas outras funções, pra ficar mais linda ainda nossa api ;D

* arametros estão em pt/br e declarados com números, acho que ficaria chique se colocassemos firstObject e secondObject hehehe;

* o return pode ser encurtado, você pode “remover” a propriedade da senha dos objetos e depois comparar objeto1 === objeto2;


```
if (email !== currentEmail || cpf !== currentCPF) {
      await existsInDatabase(req, (database = "users"));
    }
```

Fiquei com uma dúvida no trecho acima. A função existsindatabase vai retornar um thnow new error, mas se já existir esse email temos que enviar a mensagem pro cliente, essa mensagem é enviada em alguma outra função?


**CONTROLLERS** –

**costumer.js**


- POSTCOSTUMER
    * Função postCostumer deve ser declarada usando const
    
    
    * O status code deve retornar o código 201-CREATED, pois algo foi criado no banco de dados

    `return res.status(200).json("Cliente cadastrado com sucesso.");`

    * Tratar o erro para retornar a mensagem em pt/br
    
    `catch (error) {
    return res.status(400).json(error.messagem);
  }
`



**LOGIN**

* a função login deve ser declarado com const;
* const jwtPassword pode ser importada no início da função;


**POSTUSER**

- deve ser declarado com const


- deve retornar status code 201 - "Created", pois algo foi criado no banco de dados

```

return res.status(200).json("Usuário cadastrado com sucesso.");
  
```

* Tratar o erro para retornar a mensagem em pt/br


```
return res.status(400).json(error.messagem);
  }
};
```
    

**PUTUSER**

- deve ser declarado com const
- tratar erro do catch, passar para pt

 ```
const user = await connection("users")
      .where({ id })
      .update({
        name,
        email,
        cpf: currentCpf,
        telephone,
        password: passwordHash,
      })
      .returning("*");

    return res.status(201).json("Usuário atualizado com sucesso!");
    
```

como const User não foi chamada, acho que seria interessante trocar para
```
await connection("users")
      .where({ id })
      .update({
        name,
        email,
        cpf: currentCpf,
        telephone,
        password: passwordHash,
      })
      .returning("*");

    return res.status(201).json("Usuário atualizado com sucesso!");
```

Na mesma parte, em 
```
    return res.status(201).json("Usuário atualizado com sucesso!");
    ```
```

O status code retornado deve ser 204 ou 200, pois nada foi criado, somente atualizado.
Eu optaria pelo 200, mas pode ser o 204 também.


No geral tá linda nossa api, só ajustar esses detalhes para ficar ainda melhor <3. Parabéns pelo trabalho Edu!

Lembrando que não sou dev Senior, então posso ter me equivocado em algo ou não reparado. Sinta-se livre para possíveis objeções, estamos passíveis a erros o tempo todo.


**Observação:** Estou enviando nesse PR a alteração de tudo que tem que ser feito no código e não o que deve ser alterado por branchs. Não é o correto, falha minha.

eduardortm committed:
:bulb: docs: Correção das sugestões do PR

As sugestões da parte de cadastro de cliente não foram realizadas

O UserUpdate não valida mais a senha pois tem um erro que irei trabalhar

eduardortm committed:
:wrench: chore: Correção de pequenos erros

Houve uma alteração no nome de um arquivo e causava erro de importação

Criação do arquivo .env.example

eduardortm committed:
:wrench: chore: Alteração da validação de senha

Inclusão do ponto como caractere especial permitido

Correção do erro de validação da senha do schema updateUser

eduardortm committed:
:wrench: chore: implementação da validação do CEP

eduardortm committed:
:wrench: chore: Implementação de validações

Correção das validações do CustomerSchema

eduardortm committed:
:wrench: chore: Alteração de postUser

Inclusão da verificação existsIndatabase

eduardortm committed:
:bulb: Alteração da mensagem de validação de senha

eduardortm committed:
:bulb: docs: Atualização do README

Atualização do readme especificando os endpoints

Atualização de algumas verificações de e retornos de erro

- Ainda será necessário alterar algumas validações dos schemas

eduardortm committed:
:bulb: docs: atualização do README

eduardortm committed:
:bulb: docs: Atualização do README

Correção de alguns erros de digitação

eduardortm committed:
:bulb: docs: Correção de problema de importação

- Correção da impotação de user.js, alterei pra letra minúcsula

Estava com um problema pra importar User

eduardortm committed:
:bulb: docs: Correções Básicas

Correções de alguns erros de digitaçao, nomes e validações

eduardortm committed:
:bulb: docs: Correção de erro

Uma vírgula no lugar errado estava dando um erro

eduardortm committed:
:bulb: docs: inclusão da especificação da rota

Não havia especificado a rota de cadastro de clientes

Especificação de necessidade de validação para POSTcustomer

eduardortm committed:
:bulb: docs: Correção de padrão README

eduardortm committed:
:bulb: Correção de Padrão no README

eduardortm committed:
:wrench: chore: Criação da rota getUser

Criação da const getUser e implementação da rota

eduardortm committed:
:bulb: docs: Alteração do README

Atualizei o README com as informações da rota getUser

eduardortm committed:
:wrench: chore: Atualizações

Correção de algumas validações e retornos

- Nome não é mais obrigatório para cadastro de usuário
 - validação feita no front

O cpf pode estar Vazio

O Telefone Pode estar Vazio

:bulb: comment: Preciso Atualizar a documentação
